python/semantic_kernel/utils/model_diagnostics.py

```diff
@@ -0,0 +1,151 @@
+    # APPLICATIONINSIGHTS_CONNECTION_STRING environment variable.
+    connection_string = os.getenv("APPLICATIONINSIGHTS_CONNECTION_STRING")
+    if connection_string:
+        configure_azure_monitor(connection_string=connection_string)
+
+# Sets the global default tracer provider
+tracer = trace.get_tracer(__name__)
+
+
+def are_model_diagnostics_enabled() -> bool:
+    """Check if model diagnostics are enabled.
+
+    Model diagnostics are enabled if either EnableModelDiagnostics or EnableSensitiveEvents is set.
+    """
+    return _enable_diagnostics or _enable_sensitive_events
+
+
+def are_sensitive_events_enabled() -> bool:
+    """Check if sensitive events are enabled.
+
+    Sensitive events are enabled if EnableSensitiveEvents is set.
+    """
+    return _enable_sensitive_events
+
+
+def start_completion_activity(model_name: str, model_provider: str, prompt: str,
+                              execution_settings: Optional[PromptExecutionSettings]) -> Optional[Span]:
+    """Start a text completion activity for a given model."""
+    if not are_model_diagnostics_enabled():
+        return None
+
+    operation_name: str = "chat.completions" if isinstance(prompt, ChatHistory) else "text.completions"
+
+    span = tracer.start_span(f"{operation_name} {model_name}")
+
+    # Set attributes on the span
+    span.set_attributes({
+        _OPERATION: operation_name,
+        _SYSTEM: model_provider,
+        _MODEL: model_name,
+    })
+
+    if execution_settings is not None:
+        span.set_attributes({
+            _MAX_TOKEN: str(execution_settings.extension_data.get("max_tokens")),
+            _TEMPERATURE: str(execution_settings.extension_data.get("temperature")),
+            _TOP_P: str(execution_settings.extension_data.get("top_p")),
+        })
+
+    if are_sensitive_events_enabled():
+        span.add_event(PROMPT_EVENT_PROMPT, {PROMPT_EVENT_PROMPT: prompt})
+
+    return span
+
+
+def set_completion_response(span: Span, completions: list[ChatMessageContent], finish_reasons: list[str],
+                            response_id: str, prompt_tokens: Optional[int] = None,
+                            completion_tokens: Optional[int] = None) -> None:
+    """Set the text completion response for a given activity."""
+    if not are_model_diagnostics_enabled():
+        return
+
+    if prompt_tokens:
+        span.set_attribute(_PROMPT_TOKEN, prompt_tokens)
+
+    if completion_tokens:
+        span.set_attribute(_COMPLETION_TOKEN, completion_tokens)
+
+    if finish_reasons:
+        span.set_attribute(_FINISH_REASON, ",".join(finish_reasons))
+
+    span.set_attribute(_RESPONSE_ID, response_id)
+
+    if are_sensitive_events_enabled() and len(completions) > 0:
+        span.add_event(COMPLETION_EVENT_COMPLETION,
+                       {COMPLETION_EVENT_COMPLETION: _messages_to_openai_format(completions)})
+
+
+def _messages_to_openai_format(chat_history: list[ChatMessageContent]) -> str:
```

Comment:
ideally this generic function should not have openai specific functions...

Reply:
This is needed since the OTeL GenAI semantic convention recommends the OpenAI format: https://github.com/open-telemetry/semantic-conventions/blob/main/docs/gen-ai/gen-ai-spans.md#events

Reply:
This is needed since the OTeL GenAI semantic convention recommends the OpenAI format: https://github.com/open-telemetry/semantic-conventions/blob/main/docs/gen-ai/gen-ai-spans.md#events

Reply:
then let's name it `_messages_to_otel_format`!